src/chatd.cpp

```diff
@@ -4688,6 +4833,104 @@ void Chat::onUserLeave(Id userid)
     }
 }
 
+void Chat::onAddReaction(Id msgId, Id userId, std::string reaction)
+{
+    Idx messageIdx = msgIndexFromId(msgId);
+    Message *message = (messageIdx != CHATD_IDX_INVALID) ? findOrNull(messageIdx) : NULL;
+    if (message)
+    {
+        if (reaction.empty())
+        {
+            CHATID_LOG_ERROR("onAddReaction: reaction received is empty. msgid: %s", ID_CSTR(msgId));
+            return;
+        }
+
+        if (message->isManagementMessage())
+        {
+            CHATID_LOG_ERROR("onAddReaction: reaction received for a management message with msgid: %s", ID_CSTR(msgId));
+            return;
+        }
+
+        auto wptr = weakHandle();
+        mCrypto->reactionDecrypt(message, reaction)
+        .then([this, wptr, message, userId](std::shared_ptr<Buffer> data)
+        {
+            if (wptr.deleted())
+                return;
+
+            std::string reaction (data->buf(), data->bufSize());
+            message->addReaction(reaction, userId);
+
+            if (!previewMode())
```

Comment:
Can a chat in preview mode upgrade due to auto-join? In that case, does the chatroom created during the preview preserved, or is it destroyed and a new is created from scratch? If the answer if the former, then you do need to add reactions to cache.
I'll fix it

Reply:
Perhaps it's a good idea to move the addition to DB in the `Message::addReaction()` (same for `delReaction()`). TBD based on the logic of the pending-reactions queue